tests/src/Kernel/UniqueEmailTest.php

```diff
@@ -0,0 +1,586 @@
+ * @coversDefaultClass \Drupal\decoupled_auth\Plugin\Validation\Constraint\DecoupledAuthUserMailUniqueValidator
+ * @group decoupled_auth
+ */
+class UniqueEmailTest extends KernelTestBase {
+
+  use DecoupledAuthUserCreationTrait;
+  use UserCreationTrait;
+
+  /**
+   * {@inheritdoc}
+   */
+  public static $modules = ['decoupled_auth', 'user', 'system'];
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() {
+    parent::setUp();
+
+    $this->installConfig(['decoupled_auth']);
+    $this->installEntitySchema('user');
+    $this->installSchema('system', 'sequences');
+  }
+
+  /**
+   * Set the configuration for unique emails.
+   *
+   * @param string $mode
+   *   The mode. Can be one of 'all', 'none', 'include' or 'exclude'.
+   * @param array $roles
+   *   The array of role IDs if in 'include' or 'exclude' mode.
+   */
+  protected function setUniqueEmailsConfig($mode, $roles = []) {
+    $this->config('decoupled_auth.settings')
+      ->set('unique_emails.mode', $mode)
+      ->set('unique_emails.roles', $roles)
+      ->save();
+  }
+
+  /**
+   * Test the unique email validator in 'all' mode with no existing users.
+   *
+   * @covers ::validate
+   */
+  public function testModeAllNone() {
+    $this->setUniqueEmailsConfig('all');
+
+    // Test validating a decoupled user.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser();
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Decoupled user passes validation.');
+
+    // Test validating a coupled user.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(FALSE);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Coupled user passes validation.');
+  }
+
+  /**
+   * Test the unique email validator in 'all' mode with an existing decoupled
+   * user.
+   *
+   * @covers ::validate
+   */
+  public function testModeAllDecoupled() {
+    $this->setUniqueEmailsConfig('all');
+
+    $existing_user = $this->createDecoupledUser();
+
+    // Test validating a decoupled user.
+    // Expected result: Violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->email_prefix);
+    $violations = $user->validate();
+    $this->assertNotEmpty($violations, 'Decoupled user fails validation.');
+
+    // Test validating a coupled user.
+    // Expected result: Violations
+    $user = $this->createUnsavedUser(FALSE, $existing_user->email_prefix);
+    $violations = $user->validate();
+    $this->assertNotEmpty($violations, 'Coupled user fails validation.');
+  }
+
+  /**
+   * Test the unique email validator in 'all' mode with an existing coupled
+   * user.
+   *
+   * @covers ::validate
+   */
+  public function testModeAllCoupled() {
+    $this->setUniqueEmailsConfig('all');
+
+    $existing_user = $this->createUser();
+
+    // Test validating a decoupled user.
+    // Expected result: Violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->getAccountName());
```

Comment:
Decoupled users don't have `name` and the email is suffixed with `@example.com`, so `createDecoupledUser` stores the email prefix against the entity so we can retrieve it later without having to strip off the `@example.com`.